// @apollo/server/dist/esm/httpBatching.js
import { HeaderMap, runHttpQuery } from './runHttpQuery.js';
import { BadRequestError } from './internalErrorClasses.js';
export async function runBatchHttpQuery(server, batchRequest, body, contextValue, schemaDerivedData, internals) {
    const combinedResponse = {
        headers: new HeaderMap(),
        bodyChunks: null,
        completeBody: '',
    };
    const responseBodies = await Promise.all(body.map(async (bodyPiece) => {
        const singleRequest = {
            ...batchRequest,
            body: bodyPiece,
        };
        const response = await runHttpQuery(server, singleRequest, contextValue, schemaDerivedData, internals);
        if (response.completeBody === null) {
            throw Error('Incremental delivery not implemented');
        }
        for (const [key, value] of response.headers) {
            combinedResponse.headers.set(key, value);
        }
        if (response.status) {
            combinedResponse.status = response.status;
        }
        return response.completeBody;
    }));
    combinedResponse.completeBody = `[${responseBodies.join(',')}]`;
    return combinedResponse;
}
export async function runPotentiallyBatchedHttpQuery(server, httpGraphQLRequest, contextValue, schemaDerivedData, internals) {
    if (!Array.isArray(httpGraphQLRequest.body)) {
        return await runHttpQuery(server, httpGraphQLRequest, contextValue, schemaDerivedData, internals);
    }
    if (internals.allowBatchedHttpRequests) {
        return await runBatchHttpQuery(server, httpGraphQLRequest, httpGraphQLRequest.body, contextValue, schemaDerivedData, internals);
    }
    throw new BadRequestError('Operation batching disabled.');
}
